pyccel/parser/syntactic.py

```diff
@@ -1243,20 +1272,34 @@ def _visit_GeneratorExp(self, stmt):
             body = Assign(lhs, body)
 
         body.set_current_ast(parent)
-        indices = []
-        generators = list(generators)
-        while len(generators) > 0:
-            indices.append(generators[-1].target)
+
+        for loop, condition in zip(generators, conditions):
+            if condition:
+                loop.insert2body(condition)
+
+        if generators[-1].body.body:
+            generators[-1].body.body[0].blocks[0].body.insert2body(body)
+        else:
             generators[-1].insert2body(body)
-            body = generators.pop()
+
+        while(len(generators)) > 1:
+            indices.append(generators[-1].target)
+            outter_loop = generators.pop()
+            inserted_into = generators[-1]
+            if inserted_into.body.body:
+                inserted_into.body.body[0].blocks[0].body.insert2body(outter_loop)
+            else:
+                inserted_into.insert2body(outter_loop)
```

Comment:
The if block before the while loop seems to be the same as the one inside the while loop. Can you not do something like:
```python
         outer_loop = body
         while generators:
             inserted_into = generators[-1]
￼            if inserted_into.body.body:
￼                inserted_into.body.body[0].blocks[0].body.insert2body(outer_loop)
￼            else:
￼                inserted_into.insert2body(outer_loop)
￼            indices.append(generators[-1].target)
￼            outer_loop = generators.pop()

        expr = FunctionalX(outer_loop, result, lhs, indices)
```
to reduce code duplication?